src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs

```diff
@@ -81,85 +70,81 @@ public ISecretManager GetSecretManager(WebHostSettings settings)
 
         public WebScriptHostManager GetWebScriptHostManager(WebHostSettings settings)
         {
-            if (_activeHostManager != null)
-            {
-                return _activeHostManager;
-            }
-
-            lock (_syncLock)
-            {
-                EnsureInitialized(settings);
-
-                return _activeHostManager ?? _standbyHostManager;
-            }
+            return GetActiveInstance(settings, ref _activeHostManager, ref _standbyHostManager);
         }
 
         public WebHookReceiverManager GetWebHookReceiverManager(WebHostSettings settings)
         {
-            if (_activeReceiverManager != null)
-            {
-                return _activeReceiverManager;
-            }
-
-            lock (_syncLock)
-            {
-                EnsureInitialized(settings);
-
-                return _activeReceiverManager ?? _standbyReceiverManager;
-            }
+            return GetActiveInstance(settings, ref _activeReceiverManager, ref _standbyReceiverManager);
         }
 
+        /// <summary>
+        /// This method ensures that all services managed by this class are initialized
+        /// correctly taking into account specialization state transitions.
+        /// </summary>
         internal void EnsureInitialized(WebHostSettings settings)
         {
-            if (!WebScriptHostManager.InStandbyMode)
+            lock (_syncLock)
             {
-                // standby mode can only change from true to false
-                // when standby mode changes, we reset all instances
-                if (_activeHostManager == null)
+                if (!WebScriptHostManager.InStandbyMode)
                 {
-                    _settingsManager.Reset();
+                    // standby mode can only change from true to false
+                    // when standby mode changes, we reset all instances
+                    if (_activeHostManager == null)
+                    {
+                        _settingsManager.Reset();
 
-                    _activeScriptHostConfig = CreateScriptHostConfiguration(settings);
-                    _activeHostManager = new WebScriptHostManager(_activeScriptHostConfig, _secretManagerFactory, _eventManager, _settingsManager, settings);
-                    _activeReceiverManager = new WebHookReceiverManager(_activeHostManager.SecretManager);
-                    InitializeFileSystem();
+                        _activeScriptHostConfig = CreateScriptHostConfiguration(settings);
+                        _activeHostManager = new WebScriptHostManager(_activeScriptHostConfig, _secretManagerFactory, _eventManager, _settingsManager, settings);
+                        _specializationTimer?.Dispose();
```

Comment:
Once we've specialized, stop the timer. We only want it running when we're in standby mode.